models/migrations/v151.go

```diff
@@ -0,0 +1,23 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"code.gitea.io/gitea/modules/timeutil"
+	"xorm.io/xorm"
+)
+
+func addRevisionTable(x *xorm.Engine) error {
+	type Revision struct {
+		ID              int64              `xorm:"pk autoincr"`
+		Commit          string             `xorm:"NOT NULL"`
```

Comment:
I think we may need  start commit id and end commit id and is force push.

Reply:
Why is the start commit id interesting? That is easly computable form the head commit. We store the number of commits. Resolving "Commit~(NumberOfCommits - 1)" will give you that.

Reply:
But once you force push something, you cannot caculate them any more.

Reply:
Yes you can. Why do you think you can't? 

1. If you have loose on longer used commits they are not deleted immediatly, only at the next run of `git gc`
2. If you have a `ref` to something that doesn't count as "loose" by `git gc`. That is why we are creating the refs `refs/pull/%d/revision/%d` for each revision head. Oviously commits that are referenced by other commits which you keep must also be kept. So the first commit, or in fact all of the commits of each pull revision is obtainable. Using that you can also always calculate weather the new revision is a force push or not.

Reply:
@lunny ?